File: js/src/injector/ExtensionInjector.ts
import type { z } from "zod";
import type { LauncherOptions } from "../launcher/BrowserLauncher.js";
import type { TargetRoute, UpstreamTransportOptions } from "../transport/UpstreamTransport.js";
import type { cdp } from "../types/generated/cdp.js";
import type { CdpCommandSchema } from "../types/generated/zod/helpers.js";
import * as Runtime from "../types/generated/zod/Runtime.js";
import * as Target from "../types/generated/zod/Target.js";
import type { ProtocolParams, ProtocolResult } from "../types/modcdp.js";

const EXT_ID_FROM_URL = /^chrome-extension:\/\/([a-z]+)\//;
const DEFAULT_MODCDP_EXTENSION_ID = "mdedooklbnfejodmnhmkdpkaedafkehf";
const DEFAULT_MODCDP_SERVICE_WORKER_URL_SUFFIXES = ["/modcdp/service_worker.js"];
const MODCDP_READY_EXPRESSION = "Boolean(globalThis.ModCDP?.handleCommand && globalThis.ModCDP?.addCustomEvent)";
const DEFAULT_CDP_SEND_TIMEOUT_MS = 10_000;
const DEFAULT_EXECUTION_CONTEXT_TIMEOUT_MS = 10_000;
const DEFAULT_SERVICE_WORKER_PROBE_TIMEOUT_MS = 10_000;
const DEFAULT_SERVICE_WORKER_READY_TIMEOUT_MS = 60_000;
const DEFAULT_SERVICE_WORKER_POLL_INTERVAL_MS = 100;
const DEFAULT_TARGET_SESSION_POLL_INTERVAL_MS = 20;

interface SendCDP {
  (method: string, params?: ProtocolParams, session_id?: string | null): Promise<ProtocolResult>;
  <
    Params extends z.ZodType<Record<string, unknown>>,
    Result extends z.ZodType<Record<string, unknown>>,
    Name extends string,
  >(
    command: CdpCommandSchema<Params, Result, Name>,
    params?: z.input<Params>,
    route?: TargetRoute | cdp.types.ts.Target.SessionID | null,
  ): Promise<z.output<Result>>;
}
type TargetInfo = { targetId: string; type?: string; url?: string };

type InjectorMode =
  | "cli" // launch local chrome with --load-extension=/path/to/extension CLI args
  | "cdp" // connect to existing chrome and inject via Extensions.loadUnpacked(...) CDP API
  | "bb" // use Browserbase extension (via Browserbase SDK extensions upload API)
  | "discover" // auto-discover an existing extension service worker thats already running in the browser
  | "borrow" // hijack *any* extension service worker in the browser that has enough permissions to run our ModCDP server (essentially running as a parasite on some other random extension, not recommended for production)
  | "none"; // no ModCDPServer at all
type InjectorOptions = {
  injector_mode?: InjectorMode;
  send?: SendCDP | null;
  injector_cli_extension_path?: string | null;
  injector_cli_extension_id?: string | null;
  injector_cdp_extension_path?: string | null;
  injector_cdp_extension_id?: string | null;
  injector_bb_extension_path?: string | null;
  injector_bb_extension_id?: string | null;
  injector_discover_extension_path?: string | null;
  injector_borrow_extension_path?: string | null;
  injector_service_worker_extension_id?: string | null;
  injector_service_worker_url_includes?: string[];
  injector_service_worker_url_suffixes?: string[];
  injector_trust_service_worker_target?: boolean;
  injector_require_service_worker_target?: boolean;
  injector_service_worker_ready_expression?: string | null;
  injector_cdp_send_timeout_ms?: number;
  injector_execution_context_timeout_ms?: number;
  injector_service_worker_probe_timeout_ms?: number;
  injector_service_worker_ready_timeout_ms?: number;
  injector_service_worker_poll_interval_ms?: number;
  injector_target_session_poll_interval_ms?: number;
  injector_bb_api_key?: string | null;
  injector_bb_base_url?: string | null;
};

type ExtensionInjectionResult = {
  source: string;
  extension_id?: string | null;
  target_id: string;
  url?: string;
  session_id: string;
};

function delay(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

class ExtensionInjector {
  injector_mode: InjectorMode;
  send: SendCDP;
  injector_cli_extension_path: string | null;
  injector_cli_extension_id: string | null;
  injector_cdp_extension_path: string | null;
  injector_cdp_extension_id: string | null;
  injector_bb_extension_path: string | null;
  injector_bb_extension_id: string | null;
  injector_discover_extension_path: string | null;
  injector_borrow_extension_path: string | null;
  injector_service_worker_extension_id: string | null;
  injector_service_worker_url_includes: string[];
  injector_service_worker_url_suffixes: string[];
  injector_trust_service_worker_target: boolean;
  injector_require_service_worker_target: boolean;
  injector_service_worker_ready_expression: string | null;
  injector_cdp_send_timeout_ms: number;
  injector_execution_context_timeout_ms: number;
  injector_service_worker_probe_timeout_ms: number;
  injector_service_worker_ready_timeout_ms: number;
  injector_service_worker_poll_interval_ms: number;
  injector_target_session_poll_interval_ms: number;
  injector_bb_api_key: string | null;
  injector_bb_base_url: string | null;
  source: string | null;
  extension_id: string | null;
  target_id: string | null;
  url: string | null;
  session_id: string | null;
  extra_args: string[];
  protected unusable_target_ids = new Set<string>();

  constructor(options: InjectorOptions = {}) {
    this.injector_mode = options.injector_mode ?? "none";
    this.send =
      options.send ??
      (async () => {
        throw new Error(`${this.constructor.name} requires a CDP send function.`);
      });
    this.injector_cli_extension_path = options.injector_cli_extension_path ?? null;
    this.injector_cli_extension_id = options.injector_cli_extension_id ?? null;
    this.injector_cdp_extension_path = options.injector_cdp_extension_path ?? null;
    this.injector_cdp_extension_id = options.injector_cdp_extension_id ?? null;
    this.injector_bb_extension_path = options.injector_bb_extension_path ?? null;
    this.injector_bb_extension_id = options.injector_bb_extension_id ?? null;
    this.injector_discover_extension_path = options.injector_discover_extension_path ?? null;
    this.injector_borrow_extension_path = options.injector_borrow_extension_path ?? null;
    this.injector_service_worker_extension_id = options.injector_service_worker_extension_id ?? null;
    this.injector_service_worker_url_includes = options.injector_service_worker_url_includes ?? [];
    this.injector_service_worker_url_suffixes =
      options.injector_service_worker_url_suffixes ?? DEFAULT_MODCDP_SERVICE_WORKER_URL_SUFFIXES;
    this.injector_trust_service_worker_target = options.injector_trust_service_worker_target ?? false;
    this.injector_require_service_worker_target = options.injector_require_service_worker_target ?? false;
    this.injector_service_worker_ready_expression = options.injector_service_worker_ready_expression ?? null;
    this.injector_cdp_send_timeout_ms = options.injector_cdp_send_timeout_ms ?? DEFAULT_CDP_SEND_TIMEOUT_MS;
    this.injector_execution_context_timeout_ms =
      options.injector_execution_context_timeout_ms ?? DEFAULT_EXECUTION_CONTEXT_TIMEOUT_MS;
    this.injector_service_worker_probe_timeout_ms =
      options.injector_service_worker_probe_timeout_ms ?? DEFAULT_SERVICE_WORKER_PROBE_TIMEOUT_MS;
    this.injector_service_worker_ready_timeout_ms =
      options.injector_service_worker_ready_timeout_ms ?? DEFAULT_SERVICE_WORKER_READY_TIMEOUT_MS;
    this.injector_service_worker_poll_interval_ms =
      options.injector_service_worker_poll_interval_ms ?? DEFAULT_SERVICE_WORKER_POLL_INTERVAL_MS;
    this.injector_target_session_poll_interval_ms =
      options.injector_target_session_poll_interval_ms ?? DEFAULT_TARGET_SESSION_POLL_INTERVAL_MS;
    this.injector_bb_api_key = options.injector_bb_api_key ?? null;
    this.injector_bb_base_url = options.injector_bb_base_url ?? null;
    this.source = null;
    this.extension_id = null;
    this.target_id = null;
    this.url = null;
    this.session_id = null;
    this.extra_args = [];
  }

  update(config: InjectorOptions = {}) {
    this.injector_mode = config.injector_mode ?? this.injector_mode;
    this.send = config.send ?? this.send;
    this.injector_cli_extension_path = config.injector_cli_extension_path ?? this.injector_cli_extension_path;
    this.injector_cli_extension_id = config.injector_cli_extension_id ?? this.injector_cli_extension_id;
    this.injector_cdp_extension_path = config.injector_cdp_extension_path ?? this.injector_cdp_extension_path;
    this.injector_cdp_extension_id = config.injector_cdp_extension_id ?? this.injector_cdp_extension_id;
    this.injector_bb_extension_path = config.injector_bb_extension_path ?? this.injector_bb_extension_path;
    this.injector_bb_extension_id = config.injector_bb_extension_id ?? this.injector_bb_extension_id;
    this.injector_discover_extension_path =
      config.injector_discover_extension_path ?? this.injector_discover_extension_path;
    this.injector_borrow_extension_path = config.injector_borrow_extension_path ?? this.injector_borrow_extension_path;
    this.injector_service_worker_extension_id =
      config.injector_service_worker_extension_id ?? this.injector_service_worker_extension_id;
    this.injector_service_worker_url_includes =
      config.injector_service_worker_url_includes ?? this.injector_service_worker_url_includes;
    this.injector_service_worker_url_suffixes =
      config.injector_service_worker_url_suffixes ?? this.injector_service_worker_url_suffixes;
    this.injector_trust_service_worker_target =
      config.injector_trust_service_worker_target ?? this.injector_trust_service_worker_target;
    this.injector_require_service_worker_target =
      config.injector_require_service_worker_target ?? this.injector_require_service_worker_target;
    this.injector_service_worker_ready_expression =
      config.injector_service_worker_ready_expression ?? this.injector_service_worker_ready_expression;
    this.injector_cdp_send_timeout_ms = config.injector_cdp_send_timeout_ms ?? this.injector_cdp_send_timeout_ms;
    this.injector_execution_context_timeout_ms =
      config.injector_execution_context_timeout_ms ?? this.injector_execution_context_timeout_ms;
    this.injector_service_worker_probe_timeout_ms =
      config.injector_service_worker_probe_timeout_ms ?? this.injector_service_worker_probe_timeout_ms;
    this.injector_service_worker_ready_timeout_ms =
      config.injector_service_worker_ready_timeout_ms ?? this.injector_service_worker_ready_timeout_ms;
    this.injector_service_worker_poll_interval_ms =
      config.injector_service_worker_poll_interval_ms ?? this.injector_service_worker_poll_interval_ms;
    this.injector_target_session_poll_interval_ms =
      config.injector_target_session_poll_interval_ms ?? this.injector_target_session_poll_interval_ms;
    this.injector_bb_api_key = config.injector_bb_api_key ?? this.injector_bb_api_key;
    this.injector_bb_base_url = config.injector_bb_base_url ?? this.injector_bb_base_url;
    return this;
  }

  recordInjectionResult(result: ExtensionInjectionResult) {
    this.source = result.source;
    this.extension_id = result.extension_id ?? null;
    this.target_id = result.target_id;
    this.url = result.url ?? null;
    this.session_id = result.session_id;
    return this;
  }

  async prepare() {}

  async close() {}

  async inject(): Promise<ExtensionInjectionResult | null> {
    throw new Error(`${this.constructor.name}.inject is not implemented.`);
  }

  configForLauncher(): LauncherOptions {
    return {
      launcher_local_extra_args: this.extra_args,
      launcher_bb_extension_id: this.injector_bb_extension_id,
    };
  }

  configForUpstream(): UpstreamTransportOptions {
    return {};
  }

  protected readyExpression() {
    const expression = this.injector_service_worker_ready_expression;
    return expression == null || expression.length === 0
      ? MODCDP_READY_EXPRESSION
      : `(${MODCDP_READY_EXPRESSION}) && Boolean(${expression})`;
  }

  protected async targetInfos() {
    return (await this.send(Target.GetTargetsCommand, {})).targetInfos;
  }

  protected async probeTarget(target: TargetInfo): Promise<ExtensionInjectionResult | null> {
    if (this.unusable_target_ids.has(target.targetId)) return null;
    const attached = await this.send(Target.AttachToTargetCommand, {
      targetId: target.targetId,
      flatten: true,
    });
    const session_id = attached.sessionId;
    try {
      await this.send(Runtime.EnableCommand, {}, session_id);
      const probe = await this.send(
        Runtime.EvaluateCommand,
        {
          expression: this.readyExpression(),
          returnByValue: true,
        },
        session_id,
      );
      if (probe.result?.value !== true) {
        await this.send(Target.DetachFromTargetCommand, {
          sessionId: session_id,
        }).catch(() => {});
        return null;
      }
      return {
        source: "discover",
        extension_id: target.url?.match(EXT_ID_FROM_URL)?.[1],
        target_id: target.targetId,
        url: target.url,
        session_id,
      };
    } catch (error) {
      await this.send(Target.DetachFromTargetCommand, {
        sessionId: session_id,
      }).catch(() => {});
      throw error;
    }
  }

  protected async discoverReadyServiceWorker({ matched_only = false }: { matched_only?: boolean } = {}) {
    const target_infos = await this.targetInfos();
    if (this.injector_trust_service_worker_target) {
      const trusted_target = target_infos.find((candidate) => this.serviceWorkerTargetMatches(candidate)) as
        | TargetInfo
        | undefined;
      if (trusted_target) {
        const probed = await this.probeTarget(trusted_target);
        if (probed) return { ...probed, source: "trusted" };
      }
    }
    if (this.injector_trust_service_worker_target || matched_only) return null;
    for (const candidate of target_infos) {
      if (candidate.type !== "service_worker") continue;
      if (!candidate.url.startsWith("chrome-extension://")) continue;
      try {
        const probed = await this.probeTarget(candidate as TargetInfo);
        if (probed) return probed;
      } catch {
        continue;
      }
    }
    return null;
  }

  protected async waitForReadyServiceWorker(
    timeout_ms: number,
    { matched_only = false }: { matched_only?: boolean } = {},
  ) {
    const deadline = Date.now() + timeout_ms;
    while (Date.now() < deadline) {
      const discovered = await this.discoverReadyServiceWorker({
        matched_only,
      });
      if (discovered) return discovered;
      await delay(this.injector_service_worker_poll_interval_ms ?? DEFAULT_SERVICE_WORKER_POLL_INTERVAL_MS);
    }
    return null;
  }

  protected serviceWorkerTargetMatches(candidate: { type?: string; url?: string }) {
    const url = candidate.url ?? "";
    if (candidate.type !== "service_worker") return false;
    if (!url.startsWith("chrome-extension://")) return false;
    const has_extension_id = Boolean(this.injector_service_worker_extension_id);
    if (
      this.injector_service_worker_extension_id &&
      !url.startsWith(`chrome-extension://${this.injector_service_worker_extension_id}/`)
    )
      return false;
    const includes = this.injector_service_worker_url_includes ?? [];
    const suffixes = this.injector_service_worker_url_suffixes ?? [];
    if (includes.length > 0 && !includes.every((part) => url.includes(part))) return false;
    if (suffixes.length > 0 && !suffixes.some((suffix) => url.endsWith(suffix))) return false;
    return has_extension_id || includes.length > 0 || suffixes.length > 0;
  }
}

export {
  DEFAULT_MODCDP_EXTENSION_ID,
  DEFAULT_MODCDP_SERVICE_WORKER_URL_SUFFIXES,
  DEFAULT_CDP_SEND_TIMEOUT_MS,
  DEFAULT_EXECUTION_CONTEXT_TIMEOUT_MS,
  DEFAULT_SERVICE_WORKER_PROBE_TIMEOUT_MS,
  DEFAULT_SERVICE_WORKER_READY_TIMEOUT_MS,
  DEFAULT_SERVICE_WORKER_POLL_INTERVAL_MS,
  DEFAULT_TARGET_SESSION_POLL_INTERVAL_MS,
  ExtensionInjector,
};
export type { SendCDP, TargetInfo, InjectorMode, InjectorOptions, ExtensionInjectionResult };

File: python/modcdp/injector/ExtensionInjector.py
from __future__ import annotations

import base64
import hashlib
import json
import re
import shutil
import tempfile
import threading
import time
import zipfile
from collections.abc import Callable, Mapping
from pathlib import Path
from queue import Empty, Queue
from typing import Any, TypedDict, cast

from ..launcher.BrowserLauncher import LauncherOptions
from ..types.modcdp import ProtocolParams, ProtocolResult, TargetInfo

EXT_ID_FROM_URL_RE = re.compile(r"^chrome-extension://([a-z]+)/")
DEFAULT_MODCDP_EXTENSION_ID = "mdedooklbnfejodmnhmkdpkaedafkehf"
DEFAULT_MODCDP_SERVICE_WORKER_URL_SUFFIXES = ["/modcdp/service_worker.js"]
MODCDP_READY_EXPRESSION = (
    "Boolean(globalThis.ModCDP?.handleCommand && globalThis.ModCDP?.addCustomEvent)"
)
DEFAULT_CDP_SEND_TIMEOUT_MS = 10_000
DEFAULT_EXECUTION_CONTEXT_TIMEOUT_MS = 10_000
DEFAULT_SERVICE_WORKER_PROBE_TIMEOUT_MS = 10_000
DEFAULT_SERVICE_WORKER_READY_TIMEOUT_MS = 60_000
DEFAULT_SERVICE_WORKER_POLL_INTERVAL_MS = 100
DEFAULT_TARGET_SESSION_POLL_INTERVAL_MS = 20

SendCDP = Callable[[str, ProtocolParams | None, str | None], ProtocolResult]


class InjectorOptions(TypedDict, total=False):
    send: SendCDP | None
    injector_cli_extension_path: str | None
    injector_cli_extension_id: str | None
    injector_cdp_extension_path: str | None
    injector_cdp_extension_id: str | None
    injector_bb_extension_path: str | None
    injector_bb_extension_id: str | None
    injector_discover_extension_path: str | None
    injector_borrow_extension_path: str | None
    injector_service_worker_extension_id: str | None
    injector_service_worker_url_includes: list[str]
    injector_service_worker_url_suffixes: list[str]
    injector_trust_service_worker_target: bool
    injector_require_service_worker_target: bool
    injector_service_worker_ready_expression: str | None
    injector_cdp_send_timeout_ms: int
    injector_execution_context_timeout_ms: int
    injector_service_worker_probe_timeout_ms: int
    injector_service_worker_ready_timeout_ms: int
    injector_service_worker_poll_interval_ms: int
    injector_target_session_poll_interval_ms: int
    injector_bb_api_key: str | None
    injector_bb_base_url: str | None


def defaultModCDPExtensionPath() -> str | None:
    bundled_extension = Path(__file__).resolve().parent.parent / "extension.zip"
    return str(bundled_extension) if bundled_extension.exists() else None


def prepareUnpackedExtension(extension_path: str) -> tuple[str, tempfile.TemporaryDirectory[str]]:
    cleanup_dir = tempfile.TemporaryDirectory(prefix="modcdp-extension-")
    try:
        if extension_path.endswith(".zip"):
            with zipfile.ZipFile(extension_path) as archive:
                _extract_zip(archive, cleanup_dir.name)
        else:
            shutil.copytree(extension_path, cleanup_dir.name, dirs_exist_ok=True)
        return _extension_root(cleanup_dir.name), cleanup_dir
    except BaseException:
        cleanup_dir.cleanup()
        raise


def extensionIdFromManifestKey(extension_path: str) -> str | None:
    manifest_path = Path(extension_path) / "manifest.json"
    if not manifest_path.exists():
        return None
    manifest = json.loads(manifest_path.read_text())
    key = manifest.get("key") if isinstance(manifest, dict) else None
    if not isinstance(key, str) or not key.strip():
        return None
    digest = hashlib.sha256(base64.b64decode(key)).digest()[:16]
    alphabet = "abcdefghijklmnop"
    return "".join(alphabet[byte >> 4] + alphabet[byte & 0x0F] for byte in digest)


def _extension_root(unpacked_path: str) -> str:
    if (Path(unpacked_path) / "manifest.json").exists():
        return unpacked_path
    nested = Path(unpacked_path) / "extension"
    if (nested / "manifest.json").exists():
        return str(nested)
    return unpacked_path


def _extract_zip(archive: zipfile.ZipFile, destination: str) -> None:
    root = Path(destination).resolve()
    for member in archive.infolist():
        target = (root / member.filename).resolve()
        if target != root and root not in target.parents:
            raise RuntimeError(f'zip entry "{member.filename}" escapes extension extraction directory')
    archive.extractall(destination)


def _defaulted(value: Any, fallback: int) -> int:
    return fallback if value is None else int(value)


class ExtensionInjectionResult(TypedDict):
    source: str
    extension_id: str | None
    target_id: str
    url: str
    session_id: str


class ExtensionInjector:
    def __init__(self, options: InjectorOptions | None = None) -> None:
        self.options = cast(InjectorOptions, {
            "send": None,
            "injector_cli_extension_path": None,
            "injector_cli_extension_id": None,
            "injector_cdp_extension_path": None,
            "injector_cdp_extension_id": None,
            "injector_bb_extension_path": None,
            "injector_bb_extension_id": None,
            "injector_discover_extension_path": None,
            "injector_borrow_extension_path": None,
            "injector_service_worker_extension_id": None,
            "injector_service_worker_url_includes": [],
            "injector_service_worker_url_suffixes": [],
            "injector_trust_service_worker_target": False,
            "injector_require_service_worker_target": False,
            "injector_service_worker_ready_expression": None,
            "injector_cdp_send_timeout_ms": DEFAULT_CDP_SEND_TIMEOUT_MS,
            "injector_execution_context_timeout_ms": DEFAULT_EXECUTION_CONTEXT_TIMEOUT_MS,
            "injector_service_worker_probe_timeout_ms": DEFAULT_SERVICE_WORKER_PROBE_TIMEOUT_MS,
            "injector_service_worker_ready_timeout_ms": DEFAULT_SERVICE_WORKER_READY_TIMEOUT_MS,
            "injector_service_worker_poll_interval_ms": DEFAULT_SERVICE_WORKER_POLL_INTERVAL_MS,
            "injector_target_session_poll_interval_ms": DEFAULT_TARGET_SESSION_POLL_INTERVAL_MS,
            "injector_bb_api_key": None,
            "injector_bb_base_url": None,
            **dict(options or {}),
        })
        self.unusable_target_ids: set[str] = set()

    def update(self, config: InjectorOptions | None = None) -> "ExtensionInjector":
        config = cast(InjectorOptions, dict(config or {}))
        self.options = cast(
            InjectorOptions,
            {
                **self.options,
                **config,
                "injector_service_worker_url_includes": config.get(
                    "injector_service_worker_url_includes",
                    self.options.get("injector_service_worker_url_includes") or [],
                ),
                "injector_service_worker_url_suffixes": config.get(
                    "injector_service_worker_url_suffixes",
                    self.options.get("injector_service_worker_url_suffixes") or [],
                ),
            },
        )
        return self

    def configForInjector(self) -> InjectorOptions:
        return cast(InjectorOptions, dict(self.options))

    def configForLauncher(self) -> LauncherOptions:
        return {}

    def configForUpstream(self) -> dict[str, Any]:
        return {}

    def prepare(self) -> None:
        return None

    def close(self) -> None:
        return None

    def inject(self) -> ExtensionInjectionResult | None:
        raise NotImplementedError(f"{type(self).__name__}.inject is not implemented.")

    def _readyExpression(self) -> str:
        expression = self.options.get("injector_service_worker_ready_expression")
        if not expression:
            return MODCDP_READY_EXPRESSION
        return f"({MODCDP_READY_EXPRESSION}) && Boolean({expression})"

    def _sendWithTimeout(
        self,
        method: str,
        params: ProtocolParams | None = None,
        session_id: str | None = None,
        timeout_ms: int | None = None,
    ) -> ProtocolResult:
        send = self.options.get("send")
        if send is None:
            raise RuntimeError(f"{type(self).__name__} requires a CDP send function.")
        effective_timeout_ms = _defaulted(
            timeout_ms if timeout_ms is not None else self.options.get("injector_cdp_send_timeout_ms"),
            DEFAULT_CDP_SEND_TIMEOUT_MS,
        )
        if effective_timeout_ms <= 0:
            return send(method, params or {}, session_id)

        result_queue: Queue[tuple[ProtocolResult | None, BaseException | None]] = Queue(maxsize=1)

        def runSend() -> None:
            try:
                result_queue.put((send(method, params or {}, session_id), None))
            except BaseException as error:
                result_queue.put((None, error))

        threading.Thread(target=runSend, daemon=True).start()
        try:
            result, error = result_queue.get(timeout=effective_timeout_ms / 1000)
        except Empty as error:
            raise TimeoutError(f"{method} timed out after {effective_timeout_ms}ms") from error
        if error is not None:
            raise error
        return result or {}

    def _targetInfos(self) -> list[TargetInfo]:
        result = self._sendWithTimeout("Target.getTargets")
        raw_targets = result.get("targetInfos")
        if not isinstance(raw_targets, list):
            return []
        targets: list[TargetInfo] = []
        for raw_target in raw_targets:
            if not isinstance(raw_target, Mapping):
                continue
            target_id = raw_target.get("targetId")
            target_type = raw_target.get("type")
            target_url = raw_target.get("url")
            if isinstance(target_id, str) and isinstance(target_type, str) and isinstance(target_url, str):
                targets.append({"targetId": target_id, "type": target_type, "url": target_url})
        return targets

    def _probeTarget(
        self,
        target: TargetInfo,
        session_timeout_ms: int = 0,
        *,
        allow_attach: bool = False,
    ) -> ExtensionInjectionResult | None:
        target_id = target["targetId"]
        if target_id in self.unusable_target_ids:
            return None
        attached = self._sendWithTimeout(
            "Target.attachToTarget",
            {"targetId": target_id, "flatten": True},
            None,
            session_timeout_ms,
        )
        session_id = attached.get("sessionId")
        if not isinstance(session_id, str) or not session_id:
            raise RuntimeError(f"Target.attachToTarget returned no sessionId for targetId={target_id}")
        try:
            self._sendWithTimeout("Runtime.enable", {}, session_id)
            probe = self._sendWithTimeout(
                "Runtime.evaluate",
                {
                    "expression": self._readyExpression(),
                    "returnByValue": True,
                },
                session_id,
            )
            result = cast(Mapping[str, Any], probe.get("result")) if isinstance(probe.get("result"), Mapping) else {}
            value = result.get("value")
            if value is not True:
                self._sendWithTimeout("Target.detachFromTarget", {"sessionId": session_id})
                return None
            match = EXT_ID_FROM_URL_RE.match(target.get("url") or "")
            return {
                "source": "discover",
                "extension_id": match.group(1) if match else None,
                "target_id": target_id,
                "url": target["url"],
                "session_id": session_id,
            }
        except BaseException:
            self._sendWithTimeout("Target.detachFromTarget", {"sessionId": session_id})
            raise

    def _discoverReadyServiceWorker(self, *, matched_only: bool = False) -> ExtensionInjectionResult | None:
        target_infos = self._targetInfos()
        if self.options.get("injector_trust_service_worker_target"):
            for candidate in target_infos:
                if not self._serviceWorkerTargetMatches(candidate):
                    continue
                probed = self._probeTarget(
                    candidate,
                    _defaulted(self.options.get("injector_service_worker_probe_timeout_ms"), DEFAULT_SERVICE_WORKER_PROBE_TIMEOUT_MS),
                    allow_attach=True,
                )
                if probed:
                    return {**probed, "source": "trusted"}
        if self.options.get("injector_trust_service_worker_target") or matched_only:
            return None
        for candidate in target_infos:
            if candidate["type"] != "service_worker":
                continue
            if not candidate["url"].startswith("chrome-extension://"):
                continue
            try:
                probed = self._probeTarget(
                    candidate,
                    _defaulted(self.options.get("injector_service_worker_probe_timeout_ms"), DEFAULT_SERVICE_WORKER_PROBE_TIMEOUT_MS),
                )
            except Exception:
                continue
            if probed:
                return probed
        return None

    def _waitForReadyServiceWorker(self, timeout_ms: int, *, matched_only: bool = False) -> ExtensionInjectionResult | None:
        deadline = time.monotonic() + timeout_ms / 1000
        while time.monotonic() < deadline:
            discovered = self._discoverReadyServiceWorker(matched_only=matched_only)
            if discovered:
                return discovered
            time.sleep(_defaulted(self.options.get("injector_service_worker_poll_interval_ms"), DEFAULT_SERVICE_WORKER_POLL_INTERVAL_MS) / 1000)
        return None

    def _serviceWorkerTargetMatches(self, candidate: Mapping[str, object]) -> bool:
        raw_target_url = candidate.get("url")
        target_url = raw_target_url if isinstance(raw_target_url, str) else ""
        if candidate.get("type") != "service_worker":
            return False
        if not target_url.startswith("chrome-extension://"):
            return False
        extension_id = self.options.get("injector_service_worker_extension_id")
        has_extension_id = bool(extension_id)
        if extension_id and not target_url.startswith(f"chrome-extension://{extension_id}/"):
            return False
        includes = self.options.get("injector_service_worker_url_includes") or []
        suffixes = self.options.get("injector_service_worker_url_suffixes") or []
        if includes and not all(part in target_url for part in includes):
            return False
        if suffixes and not any(target_url.endswith(suffix) for suffix in suffixes):
            return False
        return bool(has_extension_id or includes or suffixes)
